drivers/mtd/mtd_config_fs.c

```diff
@@ -0,0 +1,2126 @@
+                      rc = nvs_expire_ate(fs, second_addr);
+                      if (rc < 0)
+                        {
+                          ferr("expire ate failed, addr %" PRIx32 "\n",
+                               second_addr);
+                          return rc;
+                        }
+                      break;
+                    }
+                  else
+                    {
+                      fwarn("hash conflict\n");
+                    }
+                }
+
+              if (wlk_addr == fs->ate_wra)
+                {
+                  break;
+                }
+            }
+        }
+
+      if (wlk_addr == fs->ate_wra)
+        {
+          break;
+        }
+    }
+
+end:
+  /* If the block is empty, add a gc done ate to avoid having insufficient
+   * space when doing gc.
+   */
+
+  if ((!rc) && ((fs->ate_wra & ADDR_OFFS_MASK) ==
+      (fs->geo.erasesize - 2 * sizeof(struct nvs_ate))))
+    {
+      rc = nvs_add_gc_done_ate(fs);
+    }
+
+  finfo("%" PRIu32 " Eraseblocks of %" PRIu32 " bytes\n",
+        fs->geo.neraseblocks, fs->geo.erasesize);
+  finfo("alloc wra: %" PRIu32 ", 0x%" PRIx32 "\n",
+        (fs->ate_wra >> ADDR_BLOCK_SHIFT), (fs->ate_wra & ADDR_OFFS_MASK));
+  finfo("data wra: %" PRIu32 ", 0x%" PRIx32 "\n",
+        (fs->data_wra >> ADDR_BLOCK_SHIFT), (fs->data_wra & ADDR_OFFS_MASK));
+
+  return rc;
+}
+
+/****************************************************************************
+ * Name: nvs_read_hist
+ *
+ * Description:
+ *   Read a history entry from the file system. But expired ones will return
+ *   -ENOENT.
+ *   Only when we are updating an existing kv with new data and
+ *   right before the prev one hasn't been expired, will there
+ *   be two newest entries. At that time, nvs_read_hist is used to search
+ *   for prev one.
+ *
+ * Input Parameters:
+ *   fs       - Pointer to file system.
+ *   key      - Key of the entry to be read.
+ *   key_size - Size of key.
+ *   data     - Pointer to data buffer.
+ *   len      - Number of bytes to be read.
+ *   cnt      - History counter: 0: latest entry, 1:one before latest ...
+ *   ate_addr - The addr of found ate.
+ *
+ * Returned Value:
+ *   Number of bytes read. On success, it will be equal to the number
+ *   of bytes requested to be read. When the return value is larger than the
+ *   number of bytes requested to read this indicates not all bytes were
+ *   read, and more data is available. On error returns -ERRNO code.
+ *
+ ****************************************************************************/
+
+static ssize_t nvs_read_hist(FAR struct nvs_fs *fs, FAR const uint8_t *key,
+                size_t key_size, FAR void *data, size_t len, uint16_t cnt,
+                FAR uint32_t *ate_addr)
+{
+  int rc;
+  uint32_t wlk_addr;
+  uint32_t rd_addr;
+  uint32_t hist_addr;
+  uint16_t cnt_his;
+  struct nvs_ate wlk_ate;
+  uint32_t hash_id;
+
+  cnt_his = 0;
+
+  hash_id = nvs_fnv_hash(key, key_size) % 0xfffffffd + 1;
```

Comment:
we have to avoid certain value, like 0 or 0xfffffff. 0xffffffff is  special id for close ate and gc_done ate. 0 is undefined and may be used in later version.

Reply:
If you want to resolve 0 for the future use, it's ok to keep + 1.